File: ml-benchmark/autolifter/dac/max_len_1s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = cons {Int, List} | nil Unit;

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

split = \xs: List. fix (\f: List->Int->{List, List}. \x: List. \n: Int.
  if (< n 1) then {nil unit, x}
  else match x with
    cons {h, t} -> let res = f t (- n 1) in {cons {h, res.1}, res.2}
  | _ -> {nil unit, nil unit}
  end
) xs (/ (length xs) 2);

dac = \v: List -> Int. \xs: List.
  let run = fix (\f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {_, nil _} -> xs
    | _ -> let sp = split xs in
      concat (f sp.1) (f sp.2)
    end
  ) in v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

max1s = fix (\f: Int -> Int -> List -> Int. \ma: Int. \l: Int. \xs: List.
  match xs with
    nil _ -> max ma l
  | cons {h, t} ->
    if == h 0 then f (max ma l) 0 t else f ma (+ l 1) t
  end
) 0 0;

main = dac max1s;

File: ml-benchmark/autolifter/dac/max_len_1s.ml
type lst = Cons of int * lst | Nil

let rec length x =
  match x with
  | Cons (_, t) -> length t + 1
  | Nil -> 0

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil -> y

let split xs =
  let rec f x n =
    if n < 1 then (Nil, x)
    else
      match x with
      | Cons (h, t) ->
          let res = f t (n - 1) in
          match res with
          | (a, b) -> (Cons (h, a), b)
      | _ -> (Nil, Nil)
  in
  f xs (length xs / 2)

let dac v xs =
  let rec run ys =
    match ys with
    | Nil -> ys
    | Cons (_, t) ->
        match t with
        | Nil -> ys
        | _ ->
            let sp = split ys in
            match sp with
            | (a, b) -> concat (run a) (run b)
  in
  v (run xs)

let inf = 100

let max a b = if a < b then b else a

let rec max1s_f ma l xs =
  match xs with
  | Nil -> max ma l
  | Cons (h, t) ->
      if h = 0 then max1s_f (max ma l) 0 t else max1s_f ma (l + 1) t

let max1s = max1s_f 0 0

let program = dac max1s